waiter/src/waiter/auth/oauth.clj

```diff
@@ -0,0 +1,304 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.auth.oauth
+  (:require [bidi.bidi :as bidi]
+            [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.data.codec.base64 :as b64]
+            [clojure.data.json :as json]
+            [clojure.java.io :as io]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [comb.template :as template]
+            [digest]
+            [full.async :as fa]
+            [plumbing.core :as pc]
+            [plumbing.graph :as graph]
+            [qbits.jet.client.http :as http]
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :as ring-codec]
+            [waiter.auth.authentication :as auth]
+            [waiter.cookie-support :as cookie-support]
+            [waiter.utils :as utils])
+  (:import java.security.SecureRandom
+           org.eclipse.jetty.util.HttpCookieStore$Empty
+           org.eclipse.jetty.util.UrlEncoded))
+
+(def secure-random (SecureRandom.))
```

Comment:
Can this not be a field inside `OAuthAuthenticator` and initialized explicitly in `oauth-authenticator`?

Reply:
It could.  Do we need a SecureRandom for each authenticator?

Reply:
It allows the authenticator to manage how the random numbers are generated and also helps with testability.

Reply:
I still don't follow why we should do it that way.

Reply:
To avoid dependency on global state and make the snippets more testable if we ever need to control how the random strings are generated.

Reply:
Is the resolution on this - not needed now/won't do?